### PR TITLE
PWGHF: First version of the Bs creator, selector, task

### PR DESCRIPTION
**Pull request by pchrista**

### CODE REVIEW

**Review by fcolamar — COMMENTED** [review]
Hi Panos, as a preliminary comment, let me signal that by using the doMC option approach to select the tasks to be run, it won't be possible to switch off the MC task for the HFCandidateCreatorBs and the taskBs when running on Hyperloop. This because such global options cannot be set in Hyperloop, and the defaults value 'true' will be considered.
In case you plan to run on Hyperloop and to switch off the MC tasks at some point, I would suggest to modify those codes structure moving to a single task with two processes, activated/deactivated via a PROCESS_SWITCH (that can be set as a standard configurable), look e.g. at https://github.com/AliceO2Group/O2Physics/pull/230 or https://github.com/AliceO2Group/O2Physics/pull/231

File: PWGHF/Tasks/taskBs.cxx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+      if (!(candidate.hfflag() & 1 << hf_cand_bs::DecayType::BsToDsPi)) {
+        continue;
+      }
+      if (cutYCandMax >= 0. && std::abs(YBs(candidate)) > cutYCandMax) {
+        continue;
+      }
+      auto candDs = candidate.index0_as<aod::HfCandProng3>();
+      if (std::abs(candidate.flagMCMatchRec()) == 1 << hf_cand_bs::DecayType::BsToDsPi) {
+
+        auto indexMother = RecoDecay::getMother(particlesMC, candidate.index1_as<aod::BigTracksMC>().mcParticle_as<soa::Join<aod::McParticles, aod::HfCandBsMCGen>>(), pdg::Code::kLambdaB0, true);
```
Comment (fcolamar) [inline PWGHF/Tasks/taskBs.cxx:191]:
```suggestion
        auto indexMother = RecoDecay::getMother(particlesMC, candidate.index1_as<aod::BigTracksMC>().mcParticle_as<soa::Join<aod::McParticles, aod::HfCandBsMCGen>>(), pdg::Code::kBs, true);
```
Shouldn't it be kBs?

File: PWGHF/Tasks/taskBs.cxx [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        //registry.fill(HIST("hThetaStarRecBg"), candidate.cosThetaStar(), candidate.pt());
+      }
+    } // rec
+
+    // MC gen. level
+    //Printf("MC Particles: %d", particlesMC.size());
+    for (auto& particle : particlesMC) {
+      if (std::abs(particle.flagMCMatchGen()) == 1 << hf_cand_bs::DecayType::BsToDsPi) {
+
+        auto yParticle = RecoDecay::Y(array{particle.px(), particle.py(), particle.pz()}, RecoDecay::getMassPDG(pdg::Code::kLambdaB0));
```
Comment (fcolamar) [inline PWGHF/Tasks/taskBs.cxx:239]:
```suggestion
        auto yParticle = RecoDecay::Y(array{particle.px(), particle.py(), particle.pz()}, RecoDecay::getMassPDG(pdg::Code::kBs));
```
Also here, shouldn't it be kBs?

**Review by fcolamar — CHANGES_REQUESTED** [review]
Hi Panos, I added some comments related to the taskBs.cc, cheers!

**Comment by vkucera** [comment]
Hi @pchrista , please fix the compilation warning.

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      // Printf("New gen. candidate");
+      flag = 0;
+      origin = 0;
+      // Bs → Ds+ π-
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kLambdaB0, array{int(pdg::Code::kDs), -kPiPlus}, true)) {
+        // Match Ds+ -> φπ -> K+K-π
+        auto DsCandMC = particlesMC.iteratorAt(particle.daughter0Id());
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:274]: Use `particle.daughter0_as` instead.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/Tasks/taskBs.cxx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+     {"hPtGenSig", "Bs candidates (matched);candidate #it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{300, 0., 10.}}}},
+     {"hPtGen", "MC particles (matched);candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{300, 0., 30.}}}}}};
+
+  Configurable<int> selectionFlagBs{"selectionFlagBs", 1, "Selection Flag for Bs"};
+  Configurable<double> cutYCandMax{"cutYCandMax", 0.8, "max. cand. rapidity"};
+  Configurable<std::vector<double>> bins{"pTBins", std::vector<double>{hf_cuts_bs_todspi::pTBins_v}, "pT bin limits"};
+
+  void init(o2::framework::InitContext&)
+  {
+    registry.add("hEtaGen", "MC particles (matched);#Lambda_{b}^{0} candidate #it{#eta}^{gen};entries", {HistType::kTH2F, {{100, -2., 2.}, {(std::vector<double>)bins, "#it{p}_{T} (GeV/#it{c})"}}});
```
[the remark this answers is not in the document]
Comment (pchrista) [reply PWGHF/Tasks/taskBs.cxx:129]: done

File: PWGHF/TableProducer/HFBsToDsPiCandidateSelector.cxx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    if (pTBin == -1) {
+      // Printf("Bs topol selection failed at getpTBin");
+      return false;
+    }
+
+    // check that the candidate pT is within the analysis range
+    if (candpT < pTCandMin || candpT >= pTCandMax) {
+      return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (pchrista) [reply PWGHF/TableProducer/HFBsToDsPiCandidateSelector.cxx:72]: done

**Comment by pchrista** [comment]
Hi @fcolamar @vkucera could you check and let me know what I missed?

**Comment by fcolamar** [comment]
Hi @pchrista, in the latest version of the taskBs I still see the pdg::Code::kLambdaB0 calls at L192 and L241, and two Lambda_c labels at L58 and L61.
For the rest instead the taskBs.cxx looks fine to me.

**Comment by pchrista** [comment]
Thanks @fcolamar and apologies for the delay and the omission; I just pushed some changes

**Comment by fcolamar** [comment]
Hi Panos, ok for the taskBs from my side, but of course let's see whether Vit has still comments on the other files

**Comment by pchrista** [comment]
hi @vkucera , could you give it a look? It would be great if this code is merged so that Maurice can pick it up and develop/improve it further

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 6]
```diff
@@ -1206,6 +1218,95 @@ DECLARE_SOA_TABLE(HfCandLbMCRec, "AOD", "HFCANDLBMCREC", //!
 DECLARE_SOA_TABLE(HfCandLbMCGen, "AOD", "HFCANDLBMCGEN", //!
                   hf_cand_lb::FlagMCMatchGen,
                   hf_cand_lb::OriginMCGen);
+
+// specific Bs candidate properties
+namespace hf_cand_bs
+{
+DECLARE_SOA_INDEX_COLUMN_FULL(Index0, index0, int, HfCandProng3, "_0"); // Bs index
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:1225]:
```suggestion
DECLARE_SOA_INDEX_COLUMN_FULL(Index0, index0, int, HfCandProng3, "_0"); // Ds index
```

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 7]
```diff
@@ -1206,6 +1218,95 @@ DECLARE_SOA_TABLE(HfCandLbMCRec, "AOD", "HFCANDLBMCREC", //!
...
+// MC matching result:
+DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, int8_t); // reconstruction level
+DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t); // generator level
+DECLARE_SOA_COLUMN(OriginMCRec, originMCRec, int8_t);       // particle origin, reconstruction level
+DECLARE_SOA_COLUMN(OriginMCGen, originMCGen, int8_t);       // particle origin, generator level
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:1230]: Are these needed?

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 8]
```diff
@@ -1206,6 +1218,95 @@ DECLARE_SOA_TABLE(HfCandLbMCRec, "AOD", "HFCANDLBMCREC", //!
...
+DECLARE_SOA_COLUMN(DebugMCRec, debugMCRec, int8_t);         // debug flag for mis-association reconstruction level
+// mapping of decay types
+enum DecayType { BsToDsPi }; // move this to a dedicated cascade namespace in the future?
+
+// Bs → Ds+ π- → K+ K- π+ π-
+// float massBs = RecoDecay::getMassPDG(pdg::Code::kBs);
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:1236]:
```suggestion
```

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 9]
```diff
@@ -1206,6 +1218,95 @@ DECLARE_SOA_TABLE(HfCandLbMCRec, "AOD", "HFCANDLBMCREC", //!
...
+{
+  return candidate.y(RecoDecay::getMassPDG(pdg::Code::kBs));
+}
+
+template <typename T>
+auto EBs(const T& candidate)
+{
+  return candidate.e(RecoDecay::getMassPDG(pdg::Code::kBs));
+}
+template <typename T>
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:1254]:
```suggestion
}

template <typename T>
```

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        hCovSVXX->Fill(covMatrixPCA[0]);
+        hCovPVXX->Fill(covMatrixPV[0]);
+
+        // get uncertainty of the decay length
+        double phi, theta;
+        getPointDirection(array{collision.posX(), collision.posY(), collision.posZ()}, secondaryVertexBs, phi, theta);
+        auto errorDecayLength = std::sqrt(getRotatedCovMatrixXX(covMatrixPV, phi, theta) + getRotatedCovMatrixXX(covMatrixPCA, phi, theta));
+        auto errorDecayLengthXY = std::sqrt(getRotatedCovMatrixXX(covMatrixPV, phi, 0.) + getRotatedCovMatrixXX(covMatrixPCA, phi, 0.));
+
+        int hfFlag = 1 << hf_cand_bs::DecayType::BsToDsPi;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:184]: How is this flag useful?

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 11]
```diff
@@ -1206,6 +1218,95 @@ DECLARE_SOA_TABLE(HfCandLbMCRec, "AOD", "HFCANDLBMCREC", //!
...
+{
+DECLARE_SOA_INDEX_COLUMN_FULL(Index0, index0, int, HfCandProng3, "_0"); // Bs index
+// MC matching result:
+DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, int8_t); // reconstruction level
+DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t); // generator level
+DECLARE_SOA_COLUMN(OriginMCRec, originMCRec, int8_t);       // particle origin, reconstruction level
+DECLARE_SOA_COLUMN(OriginMCGen, originMCGen, int8_t);       // particle origin, generator level
+DECLARE_SOA_COLUMN(DebugMCRec, debugMCRec, int8_t);         // debug flag for mis-association reconstruction level
+// mapping of decay types
+enum DecayType { BsToDsPi }; // move this to a dedicated cascade namespace in the future?
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:1233]:
```suggestion
enum DecayType { BsToDsPi = 0 }; // move this to a dedicated cascade namespace in the future?
```

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+  OutputObj<TH1F> hMassBsToDsPi{TH1F("hMassBsToDsPi", "2-prong candidates;inv. mass (B_{s}^{0} #rightarrow D_{s}^{#plus}#pi^{#minus} #rightarrow K^{#plus}K^{#minus}#pi^{#plus}#pi^{#minus}) (GeV/#it{c}^{2});entries", 500, 3., 8.)};
+  OutputObj<TH1F> hCovPVXX{TH1F("hCovPVXX", "2-prong candidates;XX element of cov. matrix of prim. vtx position (cm^{2});entries", 100, 0., 1.e-4)};
+  OutputObj<TH1F> hCovSVXX{TH1F("hCovSVXX", "2-prong candidates;XX element of cov. matrix of sec. vtx position (cm^{2});entries", 100, 0., 0.2)};
+
+  double massPi = RecoDecay::getMassPDG(kPiMinus);
+  double massDs = RecoDecay::getMassPDG(pdg::Code::kDs);
+  double massDsPi = 0.;
+
+  Configurable<int> d_selectionFlagDs{"d_selectionFlagDs", 1, "Selection Flag for Ds"};
+  Configurable<double> cutYCandMax{"cutYCandMax", -1., "max. cand. rapidity"};
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:68]: Keep all configurables at one place.

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      // Printf("New gen. candidate");
+      flag = 0;
+      origin = 0;
+      // Bs → Ds+ π-
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kBs, array{int(pdg::Code::kDs), -kPiPlus}, true)) {
+        // Match Ds+ -> φπ -> K+K-π
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:273]: How is this related to φπ?

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{
+    adaptAnalysisTask<HFCandidateCreatorBs>(cfgc, TaskName{"hf-cand-creator-bs"}),
+    adaptAnalysisTask<HFCandidateCreatorBsExpressions>(cfgc, TaskName{"hf-cand-creator-bs-expressions"})};
+  const bool doMC = cfgc.options().get<bool>("doMC");
+  if (doMC) {
+    workflow.push_back(adaptAnalysisTask<HFCandidateCreatorBsMC>(cfgc, TaskName{"hf-cand-creator-bs-mc"}));
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:293]: Do not use `TaskName`.

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+                                  candidate.index1_as<aod::BigTracksMC>()};
+      auto arrayDaughtersDs = array{dsCand.index0_as<aod::BigTracksMC>(),
+                                    dsCand.index1_as<aod::BigTracksMC>(),
+                                    dsCand.index2_as<aod::BigTracksMC>()};
+      // Bs → Ds+ π-
+      // Printf("Checking Bs → Ds+ π-");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kBs, array{+kKPlus, -kKPlus, +kPiPlus, -kPiPlus}, true, &sign, 2);
+      if (indexRec > -1) {
+        // Bs → Ds+ π-
+        // Printf("Checking Bs → Ds+ π-");
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:254]: Here you seem to be checking the Ds decay.

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersDs, pdg::Code::kDs, array{+kKPlus, -kKPlus, +kPiPlus}, true, &sign, 1);
+        if (indexRec > -1) {
+          flag = 1 << hf_cand_bs::DecayType::BsToDsPi;
+        } else {
+          debug = 1;
+          LOGF(info, "WARNING: Λb in decays in the expected final state but the condition on the intermediate state is not fulfilled");
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:260]: Is this an info or a warning?
What do you want to report here?

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+      int index2Ds = track2.globalIndex();
+      // int charge = track0.sign() + track1.sign() + track2.sign();
+
+      for (auto& trackPion : tracks) {
+        if (trackPion.pt() < ptPionMin) {
+          continue;
+        }
+        if (trackPion.sign() > 0) {
+          continue;
+        }
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:146]: What is the purpose of this cut?

File: PWGHF/TableProducer/HFBsToDsPiCandidateSelector.cxx (hunk 5 above)
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/TableProducer/HFBsToDsPiCandidateSelector.cxx]: I am sorry, this should not have been removed. I thought it was checking the selector array pT range, but it actually checks a configurable range.

File: PWGHF/TableProducer/HFCandidateCreatorBs.cxx [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+
+void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
+{
+  ConfigParamSpec optionDoMC{"doMC", VariantType::Bool, true, {"Perform MC matching."}};
+  workflowOptions.push_back(optionDoMC);
+}
+
+#include "Framework/runDataProcessing.h"
+
+/// Reconstruction of Λb candidates
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorBs.cxx:43]: Fix Λb

File: PWGHF/Tasks/taskBs.cxx [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+
+void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
+{
+  ConfigParamSpec optionDoMC{"doMC", VariantType::Bool, true, {"Fill MC histograms."}};
+  workflowOptions.push_back(optionDoMC);
+}
+
+#include "Framework/runDataProcessing.h"
+
+/// Λb0 analysis task
```
Comment (vkucera) [inline PWGHF/Tasks/taskBs.cxx:42]: Fix Λb

**Review by vkucera — CHANGES_REQUESTED** [review]

**Pull request closed without merging**